File: 2024/20/solution.jl
using DataStructures

f = open(joinpath(@__DIR__, "input.txt"), "r")
table = stack(split.(readlines(f), ""))
close(f)

start_block = findfirst(==("S"), table)
end_block = findfirst(==("E"), table)
directions = [CartesianIndex(dir) for dir in [(-1, 0), (1, 0), (0, -1), (0, 1)]]

function neighs(coords)
    [coords + dir for dir in directions if checkbounds(Bool, table, coords + dir)]
end

function bfs(table)
    visited = Dict{CartesianIndex,Int64}()
    to_visit = Queue{Tuple{CartesianIndex{2},Int64}}()
    enqueue!(to_visit, (start_block, 0))
    while !isempty(to_visit)
        current, cost = dequeue!(to_visit)
        if haskey(visited, current)
            continue
        end
        visited[current] = cost
        enqueue!.(Ref(to_visit), [(neigh, cost + 1) for neigh in neighs(current) if table[neigh] != "#"])
    end
    visited
end


function solution1()
    graph = bfs(table)
    path = sort(collect(graph), by=x -> x[2])
    [1 for (tile, start_cost) in path
     for dir in directions if table[tile+dir] == "#" &&
         haskey(graph, tile + dir + dir) &&
         graph[tile+dir+dir] - start_cost >= 102] |> sum
end

function solution2()
    graph = bfs(table)
    path = sort(collect(graph), by=x -> x[2])
    solution = 0
    for (start_tile, start_cost) in path
        for (end_tile, end_cost) in path[start_cost+2:end]
            distance = abs(end_tile[1] - start_tile[1]) + abs(end_tile[2] - start_tile[2])
            if distance <= 20 && end_cost - start_cost - distance >= 100
                solution += 1
            end
        end
    end
    solution
end

function solution2b()
    graph = bfs(table)
    path = sort(collect(graph), by=x -> x[2])

    [1 for (start_tile, start_cost) in path
     for distance in [CartesianIndex(x, y) for x = -20:20, y = -20:20 if abs(x) + abs(y) <= 20]
     if haskey(graph, start_tile + distance) &&
        graph[start_tile+distance] - start_cost - abs(distance[1]) - abs(distance[2]) >= 100] |> sum
end

println("solution1: ", @time solution1())
println("solution2: ", @time solution2())
println("solution2b: ", @time solution2b())

File: 2024/20/solution.rb
INPUT_FILE = __dir__ + "/input.txt"
$matrix = File.readlines(INPUT_FILE).map { |row| row.strip.split('') }
$board = nil
$height, $width = $matrix.length, $matrix[0].length

class Tile
  attr_reader :x, :y, :value
  DIRECTIONS = [[-1, 0], [0, 1], [1, 0], [0, -1]]

  def initialize(x, y, value)
    @x, @y = x, y
    @value = value
  end

  def tile_on(direction)
    $board[@x + direction[0], @y + direction[1]]
  end

  def neighs
    DIRECTIONS.map { |dx, dy| $board[x + dx, y + dy] }
  end

  def wall?
    @value == '#'
  end

end

class Board
  attr_reader :board, :start, :end_block

  def initialize
    @board = []
    $matrix.reverse.each_with_index do |row, y|
      row.each_with_index do |value, x|
        @board[x] ||= []
        @board[x][y] = Tile.new(x, y, value)
        @start = @board[x][y] if value == 'S'
        @end_block = @board[x][y] if value == 'E'
      end
    end
  end

  def neighbours(tile)
    [[1, 0], [-1, 0], [0, 1], [0, -1]].map { |dx, dy| self.[](tile.x + dx, tile.y + dy) }
  end

  def [](x, y)
    return nil if x < 0 || y < 0 || x >= $width || y >= $height
    @board[x][y]
  end

  def iterate_over_board
    (0...$width).each do |x|
      (0...$height).reverse_each do |y|
        yield self[x, y]
      end
    end
  end

  def path
    return @path if @path
    @path = {}
    to_visit = [[start, 0]]
    until to_visit.empty?
      current, cost = to_visit.shift
      next if @path.has_key?(current)
      @path[current] = cost
      to_visit += current.neighs.compact.reject { |neigh| neigh.wall? }.map { |neigh| [neigh, cost + 1] }
    end
    @path
  end

  def solution_1
    puts path[end_block]
    path.map do |tile, start_cost|
      Tile::DIRECTIONS.filter_map do |direction|
        wall_candidate = tile.tile_on(direction)
        behind_wall = wall_candidate&.tile_on(direction)
        if wall_candidate&.wall? && path.has_key?(behind_wall)
          cost = (start_cost - path[behind_wall])
          1 if cost >= 102
        end
      end.sum
    end.sum
  end

  def solution_2
    tiles = path.sort_by(&:last)
    tiles.map do |start_tile, start_cost|
      tiles[start_cost + 1..].filter_map do |end_tile, end_cost|
        distance = (end_tile.x - start_tile.x).abs + (end_tile.y - start_tile.y).abs
        1 if distance <= 20 && end_cost - start_cost - distance >= 100
      end.sum
    end.sum
  end

  def solution_2b
    path.map do |tile, start_cost|
      (-10..10).permutation(2)
    end
  end
end

$board = Board.new
puts $board.solution_1
puts $board.solution_2
